packages/leann-core/src/leann/interactive_utils.py

```diff
@@ -0,0 +1,217 @@
+
+        # Load history if exists
+        try:
+            readline.read_history_file(str(history_file))
+            readline.set_history_length(1000)
+        except FileNotFoundError:
+            pass
+
+        # Save history on exit
+        atexit.register(readline.write_history_file, str(history_file))
+
+        # Optional: Enable vi editing mode (commented out by default)
+        # readline.parse_and_bind("set editing-mode vi")
+
+        self._setup_complete = True
+
+    def handle_builtin_command(self, user_input: str) -> bool:
+        """
+        Handle built-in commands.
+
+        Args:
+            user_input: The user's input string
+
+        Returns:
+            True if command was handled, False if not a built-in command
+        """
+        if not self.enable_commands:
+            return False
+
+        command = user_input.lower().strip()
+
+        if command in ["quit", "exit", "q"]:
+            print("Goodbye!")
+            return "quit"
+
+        elif command == "help":
+            self._show_help()
+            return True
+
+        elif command == "clear":
+            os.system("clear" if os.name != "nt" else "cls")
+            return True
+
+        elif command == "history":
+            self._show_history()
+            return True
+
+        # Check custom commands
+        for cmd_name, handler in self.custom_commands.items():
+            if command == cmd_name.lower():
+                return handler(user_input)
+
+        return False
+
+    def _show_help(self):
+        """Show available commands."""
+        print("Commands:")
+        print("  quit/exit/q - Exit the chat")
+        print("  help - Show this help message")
+        print("  clear - Clear screen")
+        print("  history - Show command history")
+
+        # Show custom commands
+        if self.custom_commands:
+            for cmd_name in self.custom_commands.keys():
+                print(f"  {cmd_name} - Custom command")
+
+    def _show_history(self):
+        """Show command history."""
+        history_length = readline.get_current_history_length()
+        if history_length == 0:
+            print("  No history available")
+            return
+
+        for i in range(history_length):
+            item = readline.get_history_item(i + 1)
+            if item:
+                print(f"  {i + 1}: {item}")
+
+    def get_user_input(self) -> Optional[str]:
+        """
+        Get user input with readline support and command handling.
+
+        Returns:
+            User input string, None if quit command, or continues loop for built-in commands
+        """
+        try:
+            user_input = input(self.prompt).strip()
+
+            # Handle built-in commands
+            result = self.handle_builtin_command(user_input)
+            if result == "quit":
+                return None
+            elif result is True:  # Command was handled, continue loop
+                return self.get_user_input()
```

Comment:
This function recurses whenever you run a built-in command (help/clear/history) or hit Ctrl+C. Typing help a few hundred times will eventually hit the recursion limit. Maybe we can return a sentinel from `handle_builtin_command` and loop in `run_interactive_loop` instead of recursing.

A simple reproducible code could be:
```python
import sys
sys.setrecursionlimit(100)

def f(i):
    if i > 0:
        return f(i-1)
    return 0

try:
    f(200)
except RecursionError:
    print('RecursionError encountered')
```